extensions/parquet/base/src/main/java/io/deephaven/parquet/base/RowGroupReaderImpl.java

```diff
@@ -34,40 +41,98 @@ final class RowGroupReaderImpl implements RowGroupReader {
+        // makes it harder to keep the two in-sync.
+        final Set<String> nonUniqueKeys = new HashSet<>();
+        final Set<Integer> nonUniqueFieldIds = new HashSet<>();
+        final Iterator<Type> fieldsIt = schema.getFields().iterator();
+        final Iterator<ColumnChunk> colsIt = rowGroup.getColumnsIterator();
+        while (fieldsIt.hasNext() && colsIt.hasNext()) {
+            final Type ft = fieldsIt.next();
+            final ColumnChunk column = colsIt.next();
+            final List<String> path_in_schema = column.getMeta_data().path_in_schema;
+            final String key = path_in_schema.toString();
+            final List<Type> nonRequiredFields = new ArrayList<>();
             for (int indexInPath = 0; indexInPath < path_in_schema.size(); indexInPath++) {
                 Type fieldType = schema
                         .getType(path_in_schema.subList(0, indexInPath + 1).toArray(new String[0]));
-                if (fieldType.getRepetition() != Type.Repetition.REQUIRED) {
+                if (fieldType.getRepetition() != Repetition.REQUIRED) {
                     nonRequiredFields.add(fieldType);
                 }
             }
-            schemaMap.put(key, nonRequiredFields);
+            if (chunkMap.putIfAbsent(key, column) != null) {
+                nonUniqueKeys.add(key);
+            }
+            schemaMap.putIfAbsent(key, nonRequiredFields);
+            if (ft.getId() != null) {
+                final int fieldId = ft.getId().intValue();
+                if (chunkMapByFieldId.putIfAbsent(fieldId, column) != null) {
+                    nonUniqueFieldIds.add(fieldId);
+                }
+                schemaMapByFieldId.putIfAbsent(fieldId, nonRequiredFields);
+            }
+        }
+        if (fieldsIt.hasNext() || colsIt.hasNext()) {
+            throw new IllegalStateException(String.format("Unexpected, iterators not exhausted, rootURI=%s", rootURI));
+        }
+        for (String nonUniqueKey : nonUniqueKeys) {
+            chunkMap.remove(nonUniqueKey);
+            schemaMap.remove(nonUniqueKey);
+        }
+        for (Integer nonUniqueFieldId : nonUniqueFieldIds) {
+            chunkMapByFieldId.remove(nonUniqueFieldId);
+            schemaMapByFieldId.remove(nonUniqueFieldId);
         }
         this.version = version;
     }
 
     @Override
     @Nullable
-    public ColumnChunkReaderImpl getColumnChunk(@NotNull final String columnName, @NotNull final List<String> path) {
-        final String key = path.toString();
-        final ColumnChunk columnChunk = chunkMap.get(key);
-        final List<Type> fieldTypes = schemaMap.get(key);
-        if (columnChunk == null) {
-            return null;
+    public ColumnChunkReaderImpl getColumnChunk(@NotNull final String columnName, @NotNull final List<String> path,
+            @Nullable Integer fieldId) {
+        final ColumnChunk columnChunk;
+        final List<Type> nonRequiredFields;
+        PARTS: {
```

Comment:
Does the additional code-complexity of using non-final fields warrant the cognitive burden of introducing a labeled break?